### PR TITLE
Issue #3333: add new literals in WhiteSpaceAfterCheck

### PR DESCRIPTION
**Pull request by MaksimP**
Issue #3333 
Add new tokens in WhiteSpaceAfterCheck.
LITERAL_DO, LITERAL_WHILE, LITERAL_FOR, LITERAL_IF, LITERAL_ELSE, DO_WHILE.

### CODE REVIEW

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAfterCheckTest.java [hunk 1]
```diff
@@ -65,20 +65,70 @@ public void testDefault() throws Exception {
 
     @Test
     public void testCast() throws Exception {
+        final DefaultConfiguration configurationForTestCast =
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAfterCheckTest.java]: please name it configurationTypeCast, please do the same approach for other naming.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputIfElse.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    void testIfElse() {
+        //Valid
+        if (condition()) {
+            testIfElse();
+        } else {
+            testIfElse();
+        }
+
+        //Invalid
+        if(condition()) {
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputIfElse.java]: please put "//violation" as training comment. to easily catch lines where you expect violation.
Please do the same for other input code.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAfterCheckTest.java [hunk 3]
```diff
@@ -65,20 +65,70 @@ public void testDefault() throws Exception {
...
+            "27:15: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "else"),
+        };
+        verify(configurationForTestLiteralElse, getPath("InputIfElse.java"), expected);
+    }
+
+    @Test
+    public void testLiteralFor() throws Exception {
+        final DefaultConfiguration configurationForTestLiteralFor =
+                createCheckConfig(WhitespaceAfterCheck.class);
+        configurationForTestLiteralFor.addAttribute("tokens", "LITERAL_FOR");
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAfterCheckTest.java]: please make separate Input file to test FOR, WHILE and all other new tokens.
Re-usage of input files is ok, but it complicate contribution to next developer. It is easier to debug simple Input files with comments like "//violation" on lines where violation is expected.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAfterCheckTest.java [hunk 4]
```diff
@@ -65,20 +65,70 @@ public void testDefault() throws Exception {
...
+
+    @Test
+    public void testLiteralFor() throws Exception {
+        final DefaultConfiguration configurationTestLiteralFor =
+                createCheckConfig(WhitespaceAfterCheck.class);
+        configurationTestLiteralFor.addAttribute("tokens", "LITERAL_FOR");
+        final String[] expected = {
+            "51:12: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "for"),
+        };
+        verify(configurationTestLiteralFor, getPath("InputWhitespaceAfter.java"), expected);
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAfterCheckTest.java]: I see tests for `while`, `if`, `else`, `for` but not for `do`. Can you please add something for `do`.

**Comment by rnveach** [comment]
@MaksimP Code should remain as 1 commit when merging finalized code.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAfterCheckTest.java [hunk 5]
```diff
@@ -132,6 +132,28 @@ public void testLiteralFor() throws Exception {
...
+        verify(configurationTestLiteralDo, getPath("InputWhitespaceAfter.java"), expected);
+    }
+
+    @Test
+    public void testDoWhile() throws Exception{
+        final DefaultConfiguration configurationTestDoWhile =
+                createCheckConfig(WhitespaceAfterCheck.class);
+        configurationTestDoWhile.addAttribute("tokens", "DO_WHILE");
+        final String[] expected = {
+            "18:16 " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "do{}while"),
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAfterCheckTest.java]: @MaksimP , what is "do{}while" ? you expect space after "while" (the fact that it is do-while is known only to AST tree), but violation is on "while" token.

Comment (MaksimP) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAfterCheckTest.java]: yes, i was doing whith "while" too, but no working

**Comment by romani** [comment]
Please squash all commits in one, reply to all review comments as "done", rebase commit on latest master code.

**Comment by rnveach** [comment]
@romani commits were squashed, and it looks like all previous issues were addressed.

File: src/xdocs/config_whitespace.xml [hunk 6]
```diff
@@ -1884,7 +1884,19 @@ public long toMicros(long d) { return d / (C1 / C0); }
...
+                <a
+                href="apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_ELSE">LITERAL_ELSE</a>,
+                <a
+                href="apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_WHILE">LITERAL_WHILE</a>,
+                <a
+                href="apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_DO">LITERAL_DO</a>,
+                <a
+                href="apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_FOR">LITERAL_FOR</a>,
+                <a
+                href="apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_DO">DO_WHILE</a>.
```
Comment (rnveach) [inline src/xdocs/config_whitespace.xml]: it is LITERAL_DO instead of DO_WHILE. Please fix.

File: src/xdocs/config_whitespace.xml [hunk 7]
```diff
@@ -1893,7 +1905,19 @@ public long toMicros(long d) { return d / (C1 / C0); }
...
```
Comment (rnveach) [inline src/xdocs/config_whitespace.xml]: Same here.

**Comment by romani** [comment]
@MaksimP , please reply as "DONE" to each raised point. It is not clear that your code is ready for review again.

File: src/xdocs/config_whitespace.xml (hunk 6 above)
Comment (MaksimP) [inline src/xdocs/config_whitespace.xml]: done

File: src/xdocs/config_whitespace.xml (hunk 7 above)
Comment (MaksimP) [inline src/xdocs/config_whitespace.xml]: done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputDoWhile.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+        return false;
+    }
+
+    void testDoWhile() {
+        //Valid
+        do {
+            testDoWhile();
+        } while (condition());
+
+        //Invalid
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputDoWhile.java]: please remove this comment, and all other "//invalid".

**Comment by romani** [comment]
@MaksimP , ping.

**Comment by romani** [comment]
there still changes in src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java , if you do not know how to avoid them , please redo PR.

**Comment by MaksimP** [comment]
https://maksimp.github.io/index.html

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputDoWhile.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
+package com.puppycrawl.tools.checkstyle.checks.whitespace;
+
+public class InputDoWhile {
+
+    boolean condition() {
+        return false;
+    }
+
+    void testDoWhile() {
+        //Valid
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputDoWhile.java]: please remove such comment in all Input files. You do violation on certain line , not a structure.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputDoWhile.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
+package com.puppycrawl.tools.checkstyle.checks.whitespace;
+
+public class InputDoWhile {
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputDoWhile.java]: please double all test blocks(valid and with violation) to make new blocks as single line. 
Example:  "do { smth(); } while (condition)".

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputWhitespaceAfter.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
+package com.puppycrawl.tools.checkstyle.checks.whitespace;
+
+public class InputWhitespaceAfter {
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputWhitespaceAfter.java]: please double all test blocks(valid and with violation) to make new blocks as single line. 
Example:  "do { smth(); } while (condition)".

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputDoWhile.java (hunk 10 above)
Comment (MaksimP) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputDoWhile.java]: done

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputWhitespaceAfter.java (hunk 11 above)
Comment (MaksimP) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputWhitespaceAfter.java]: done

**Comment by romani** [comment]
ok , ready to merge if stress testing on all projects of checkstyle-tester is pass. @MaksimP , please provide report.

**Comment by rnveach** [comment]
@romani I noticed that the diff for files in this PR are showing what looks like a chmod change.
Example: https://github.com/checkstyle/checkstyle/pull/3524/files#diff-615e131a790325af0f067fd848adfee9
> src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAfterCheck.java 100644 → 100755 

Does this matter? Should all files be a specific chmod?

**Comment by romani** [comment]
@MaksimP , please restore mod of files.

FYI: #3600

**Comment by MaksimP** [comment]
update report on the site https://maksimp.github.io/
add another one report

**Comment by romani** [comment]
this report tell us nothing, looks like you used  https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/checks-nonjavadoc-error.xml#L419
we do not need to test all other Checks, it is just waster of time and rubbish in report. Please use https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/my_check.xml#L23 and put here only your Check, as all tokens are default you can skip defining them.
Make `severity=error` at https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/my_check.xml#L11 it will skip all violations reporting and print only exceptions :) if any .
Please uncomment here all projects - https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/projects-to-test-on.properties to make real testing on a lot of projects.
Launch.
Share result, ideally there should be no errors ( no exceptions are expected :) ).

**Comment by romani** [comment]
@MaksimP , do you need help with checkstyle-tester ?

**Comment by MaksimP** [comment]
I'm sorry, I was distracted for a couple of days, and only began to make
No, I do not need help
or have one question. When "severity=error" the maven-site-plugin can't generate site.

**Comment by romani** [comment]
> When "severity=error" the maven-site-plugin can't generate site

yes, that is expected. Please make your Check to have "warning" level, in this case all errors will be Exceptions from Check (each case that caused exception should be moved to UT and fixed).

**Comment by MaksimP** [comment]
update report.

**Comment by rnveach** [comment]
@MaksimP Your report is missing the `xref` folder and we cannot see the source violations appear for.

Example: click on the line number for the first violation and https://maksimp.github.io/xref/pmd/pmd-core/src/main/java/net/sourceforge/pmd/cli/PMDParameters.html#L137 gives a 404.

**Comment by MaksimP** [comment]
fix error

**Comment by romani** [comment]
@MaksimP ,

Thanks lot , report look good.
But I requested testing on all projects that we have in property file. Pleaese provide.

**Comment by MaksimP** [comment]
all in one report ?

**Comment by MaksimP** [comment]
add new projects in the report

**Comment by romani** [comment]
@MaksimP , thanks a lot .
Impressive amount of badly formatted code was found.

Your report is also found that one of our CI does not work :( - wercker. It suppose to catch regression in Orekit and other project but it did not start.
Wercker config - https://github.com/checkstyle/checkstyle/blob/master/wercker.yml
Projects that should not have any error: 
Apache Apex - no report
XWiki - no report
Orekit - https://maksimp.github.io/Orekit/ - all violations are valid. 


Please do: 
1) open PR on Orekit project and propose fix for this violations.
2) generate reports for xwiki and Apex and share with us.

Unfortunately by our CI we confirm that there is not violations at all in that three projects. If I just merge your PR, our CI will fail and we start loosing control of changes in checkstyle project.

**Comment by romani** [comment]
@MaksimP , looks like wercker hooks become to work, please rebase your changes on latest master  and we will see violation logs for all that 3 projects.

**Comment by MaksimP** [comment]
(Apache Apex and XWiki) they are not in projects-to-test-on.properties. 
I added Apache Apex v3.2.1 in the report. At first broke by reason (#set( $symbol_pound = '#' ) in java files  https://github.com/apache/apex-core/blob/master/apex-app-archetype/src/main/resources/archetype-resources/src/main/java/__packageInPathFormat__/Application.java  ,  https://github.com/apache/apex-core/blob/master/apex-app-archetype/src/main/resources/archetype-resources/src/main/java/__packageInPathFormat__/RandomNumberGenerator.java  and (import ${package}.Application;)  in  https://github.com/apache/apex-core/blob/master/apex-app-archetype/src/main/resources/archetype-resources/src/test/java/__packageInPathFormat__/ApplicationTest.java  .
I commented out for report.
Which one of XWiki do ?

**Comment by MaksimP** [comment]
very much violations in OpenJDK and Findbugs

**Comment by romani** [comment]
@MaximP , please look at wercker.yml  exact clone commands that you need to do.
You can use whole command to do validation.

**Comment by MaksimP** [comment]
fixed violations in Orekit projects
update orekit report

**Comment by romani** [comment]
@MaksimP , thanks a lot.
I upgraded wercker to use latest Orekit code - #3626 .
Please do rebase your PR (to get this change) to let Wercker pass or fail on next validation.

**Comment by romani** [comment]
@MaksimP , thanks a lot ! you do awesome job !
sorry that merge was not quick, it is not typical PR , but some time such complexity happens.

**Pull request merged**